File: src/Utils/Logger.ts
export class Logger {
    constructor(show_msg: boolean,
                show_wrn: boolean,
                show_err: boolean,
                autosave: boolean,
                write_msg: boolean,
                write_wrn: boolean,
                write_err: boolean,
                compile_log_path: string,
                runtime_log_path: string){

        Logger.instances_count++;

        this.show_msg = show_msg
        this.show_wrn = show_wrn
        this.show_err = show_err
        this.autosave = autosave
        this.write_msg = write_msg
        this.write_wrn = write_wrn
        this.write_err = write_err
        this.build_log_path = compile_log_path
        this.runtime_log_path = runtime_log_path

        this.build_file = IsGame() ? undefined : io.open(compile_log_path, "w") as [LuaFile]
    };

    public msg(text: string){
        if (!this.show_msg && !this.write_msg){
            return
        }

        let msg = this.format('msg', text)
        if (this.show_msg){
            print(msg)
        }
        if (this.write_msg){
            this.log.push(msg)
            if (this.autosave){
                this.save()
            }
        }
    }

    public wrn(text: string){
        if (!this.show_wrn && !this.write_wrn){
            return
        }

        let msg = this.format('Wrn', text)
        if (this.show_wrn){
            print(msg)
        }
        if (this.write_wrn){
            this.log.push(msg)
            if (this.autosave){
                this.save()
            }
        }
    }

    public err(text: string, lvl?: number | undefined): never{
        let msg = this.format('ERROR', text)
        
        if (this.show_err){
            //print(msg)
        }
        if (this.write_err){
            this.log.push(msg)
            if (this.autosave){
                this.save()
            }
        }

        error(msg, lvl)
    }

    public save(){
        if (IsGame()){
            PreloadGenClear()
            PreloadGenStart()

            let name = Logger.runtime_file_name + Logger.instances_count.toString() + '.txt'
            Preload("\")\r\n\tDEL " + this.runtime_log_path + "\r\n\t(\"")
            Preload("\")\r\n\tDEL " + name + "\r\n\t(\"")
            for (let i = 0; i < this.log.length; i++){
                let bat_msg = Logger.fmt("\")\r\n\techo %s >> %s \r\n\t(\"",
                                         this.log[i],
                                         this.runtime_log_path)
                Preload(bat_msg)
            }
            Preload("\")\r\n\tstart " + this.runtime_log_path + "\r\n\t(\"")
            PreloadGenEnd(name)
        } else {
            if (this.build_file){
                for (this.saved; this.saved < this.log.length; this.saved++) {
                    this.build_file[0].write(this.log[this.saved] + '\n');
                }
                this.build_file[0].flush()
            }
        }
    }

    private format(pref: string, text:string){
        return Logger.fmt("[%.3f] {%s} %s",
                          os.clock() - Logger.start_time,
                          pref, text)
    }

    private static instances_count = 0;
    private static fmt = string.format;
    private static start_time = os.clock();
    private static runtime_file_name = 'rename_to_bat_and_run'

    readonly show_msg:boolean;
    readonly show_wrn:boolean;
    readonly show_err:boolean;
    readonly autosave:boolean;
    readonly write_msg:boolean;
    readonly write_wrn:boolean;
    readonly write_err:boolean;
    readonly build_log_path:string;
    readonly runtime_log_path:string;

    private log: string[] = [];
    private saved: number = 0;
    private build_file: [LuaFile] | undefined;
}

export namespace Logger {
    let sep = IsGame() ? '\\' : _G.package.config.charAt(0);
    export let Default = new Logger(true, true, true,
                                    true,
                                    true, true, true,
                                    GetDst() + sep + '..' + sep + 'log.txt',
                                    'log.txt')
}

File: src/Wc3Utils/Logger.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass')
local fmt = string.format or error('')

--=======
-- Class
--=======

local Logger = Class.new('Logger')
---@class Logger
local public = Logger.public
---@class LoggerClass
local static = Logger.static
---@type LoggerClass
local override = Logger.override
local private = {}

--=========
-- Static
--=========

---@param show_msg boolean
---@param show_wrn boolean
---@param show_err boolean
---@param autosave boolean
---@param write_msg boolean
---@param write_wrn boolean
---@param write_err boolean
---@param compile_log_path string
---@param runtime_log_path string
---@return Logger
function static.new(show_msg, show_wrn, show_err,
                    autosave, write_msg, write_wrn, write_err,
                    compile_log_path, runtime_log_path)

    local instance = Class.allocate(Logger)
    private.newData(instance,
                    show_msg, show_wrn, show_err,
                    autosave, write_msg, write_wrn, write_err,
                    compile_log_path, runtime_log_path)

    return instance
end

--========
-- Public
--========

---@param text string
function public:msg(text)
    local priv = private.data[self]

    local log_text = fmt("[%.3f] {%s} %s", os.clock() - private.start_time, 'Msg', text)
    if priv.show_msg then
        print(log_text)
    end

    if priv.write_msg then
        table.insert(priv.full_log, log_text)
        if priv.autosave then
            self:save()
        end
    end
end

---@param text string
function public:wrn(text)
    local priv = private.data[self]

    local log_text = fmt("[%.3f] {%s} %s", os.clock() - private.start_time, 'Wrn', text)
    if priv.show_wrn then
        print(log_text)
    end

    if priv.write_wrn then
        table.insert(priv.full_log, log_text)
        if priv.autosave then
            self:save()
        end
    end
end

---@param text string
---@param level number
function public:err(text, level)
    local priv = private.data[self]

    local log_text = fmt("[%.3f] {%s} %s", os.clock() - private.start_time, 'Err', text)
    if priv.show_err then
        print(log_text)
    end

    if priv.write_err then
        table.insert(priv.full_log, log_text)
        if priv.autosave then
            self:save()
        end
    end
    error('', (level or 1) + 1)
end

function public:save()
    local priv = private.data[self]

    if not IsGame() then
        for i = priv.saved + 1, #priv.full_log do
            priv.file:write(priv.full_log[i])
        end
        priv.saved = #priv.full_log
        priv.file:flush()
    else
        PreloadGenClear()
        PreloadGenStart()

        Preload("\")\r\n\tDEL "..priv.runtime_log_path.."\r\n\t(\"")
        Preload("\")\r\n\tDEL "..private.runtime_file_name.."\r\n\t(\"")
        for i = 1, #priv.full_log do
            local bat_msg = fmt("\")\r\n\techo %s >> %s \r\n\t(\"", priv.full_log[i], priv.runtime_log_path)
            Preload(bat_msg)
        end
        Preload("\")\r\n\tstart "..priv.runtime_log_path.."\r\n\t(\"")
        PreloadGenEnd(private.runtime_file_name)

        priv.saved = #priv.full_log
    end
end

--=========
-- Private
--=========

private.data = setmetatable({}, {__mode = 'k'})

private.start_time = os.clock()
private.runtime_file_name = 'rename_to_bat_and_run.txt'

---@param self Logger
---@param show_msg boolean
---@param show_wrn boolean
---@param show_err boolean
---@param autosave boolean
---@param write_msg boolean
---@param write_wrn boolean
---@param write_err boolean
---@param compile_log_path string
---@param runtime_log_path string
function private.newData(self,
                         show_msg, show_wrn, show_err,
                         autosave, write_msg, write_wrn, write_err,
                         compile_log_path, runtime_log_path)
    local priv = {
        show_msg = show_msg,
        show_wrn = show_wrn,
        show_err = show_err,

        autosave = autosave,
        write_msg = write_msg,
        write_wrn = write_wrn,
        write_err = write_err,

        compile_log_path = compile_log_path,
        runtime_log_path = runtime_log_path,

        full_log = {},
        saved = 0
    }

    if not IsGame() then
        priv.file = assert(io.open(compile_log_path, "w"))
    end

    private.data[self] = priv
end

return static